test/test-try-commit.js

```diff
@@ -0,0 +1,523 @@
+				t.ok(res.errors);
+				t.is(res.errors.length, 1);
+				const error = res.errors[0];
+				t.match(error.message, /Test failed via `t\.fail\(\)`/);
+				t.is(error.name, 'AssertionError');
+				res.commit();
+			})
+			.then(() => {
+				a.end();
+			});
+	}).run().then(result => {
+		t.true(result.passed);
+	});
+});
+
+test('try-commit does not allow to use .end() in attempt when parent is callback test', t => {
+	return ava.cb(a => {
+		a
+			.try(b => {
+				b.pass();
+				b.end();
+			})
+			.then(res => {
+				res.commit();
+				a.end();
+			});
+	}).run().then(result => {
+		t.false(result.passed);
+		t.ok(result.error);
+		t.match(result.error.message, /Error thrown in test/);
+		t.is(result.error.name, 'AssertionError');
+		t.match(result.error.values[0].formatted, /t\.end.*not supported/);
+	});
+});
+
+test('try-commit can be discarded', t => {
+	const instance = ava(a => {
+		const p = a.try(b => {
+			return new Promise(resolve => setTimeout(resolve, 500))
+				.then(() => b.pass());
+		});
+
+		p.discard();
+
+		return p.then(res => {
+			t.is(res, null);
+		});
+	});
+
+	return instance.run().then(result => {
+		t.false(result.passed);
+		t.is(instance.assertCount, 0);
+	});
+});
+
+test('try-commit accepts macros', t => {
+	const macro = b => {
+		t.is(b.title, ' Title');
+		b.pass();
+	};
+
+	macro.title = providedTitle => `${providedTitle ? providedTitle : ''} Title`;
+
+	return ava(a => {
+		return a
+			.try(macro)
+			.then(res => {
+				t.true(res.passed);
+				res.commit();
+			});
+	}).run().then(result => {
+		t.true(result.passed);
+	});
+});
+
+test('try-commit accepts multiple macros', t => {
+	const macros = [b => b.pass(), b => b.fail()];
+	return ava(a => {
+		return a.try(macros)
+			.then(([res1, res2]) => {
+				t.true(res1.passed);
+				res1.commit();
+				t.false(res2.passed);
+				res2.discard();
+			});
+	}).run().then(result => {
+		t.true(result.passed);
+	});
+});
+
+test('try-commit returns results in the same shape as when implementations are passed', t => {
+	return ava(a => {
+		return Promise.all([
+			a.try(b => b.pass()).then(results => {
+				t.match(results, {passed: true});
+				results.commit();
+			}),
+			a.try([b => b.pass()]).then(results => {
+				t.is(results.length, 1);
+				t.match(results, [{passed: true}]);
+				results[0].commit();
+			}),
+			a.try([b => b.pass(), b => b.fail()]).then(results => {
+				t.is(results.length, 2);
+				t.match(results, [{passed: true}, {passed: false}]);
+				results[0].commit();
+				results[1].discard();
+			})
+		]);
+	}).run().then(result => {
+		t.true(result.passed);
+	});
+});
+
+test('try-commit abides timeout', t => {
+	return ava(a => {
+		a.timeout(10);
+		return a.try(b => {
+			b.pass();
+			return delay(200);
+		}).then(result => result.commit());
+	}).run().then(result => {
+		t.is(result.passed, false);
+		t.match(result.error.message, /timeout/);
+	});
+});
+
+test('try-commit refreshes the timeout on commit/discard', t => {
```

Comment:
Should we test that it does not reset when assertions are run within the attempt?

Reply:
I've come up with the test, but it's hard to reason if the test is correct.